### ISSUE TITLE
add c++ test, embeddabilty

### ISSUE THREAD

**Issue by rurban** [perl11/cperl#227]
perl.h, EXTERN.h and XSUB.h needs to be embeddable into C++ projects.
add a test `t/porting/embedcpp.t` and try to find a C++ compiler on the system to compile

```
#include "EXTERN.h"
#include "perl.h"
#include "XSUB.h"

#include <iostream>
using namespace std;
int main() {
  cout << "ok 1 - from c++\n";
  return 0;
}
``` 

via `$CXX -c -I.. -Wall tmp.cc`.

# findings

## darwin
darwin clashes with our do_open, do_close macros in their iostream _OutputIterator.
Same on freebsd
```
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/locale:3663:21: error: too few arguments provided to function-like macro invocation
        do_close(__c);
                    ^
../embed.h:125:9: note: macro 'do_close' defined here
#define do_close(a,b)           Perl_do_close(aTHX_ a,b)
```

```
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/locale:3672:69: error: too few arguments provided to function-like macro invocation
    virtual catalog do_open(const basic_string<char>&, const locale&) const;
                                                                    ^
../perl.h:6934:9: note: macro 'do_open' defined here
#define do_open(g, n, l, a, rm, rp, sf)
```

undefine them when used with darwin /usr/bin/c++
rename them to be used with a mandatory Perl_ prefix. 
note that do_open is a mathom API and needs to be deprecated.
used in lib/ExtUtils/typemap, dist/Time-HiRes/typemap, dist/IO/IO.xs

## msvc

[00:08:00] # c:\Program Files (x86)\Microsoft Visual Studio 10.0\VC\INCLUDE\limits(378) : error C2766: explicit specialization; 'std::numeric_limits<char>' has already been defined
[00:08:00] #         c:\Program Files (x86)\Microsoft Visual Studio 10.0\VC\INCLUDE\limits(187) : see previous definition of 'numeric_limits<char>'
[00:08:00] # c:\Program Files (x86)\Microsoft Visual Studio 10.0\VC\INCLUDE\xtr1common(244) : error C2766: explicit specialization; 'std::tr1::_Is_integral<char>' has already been defined
[00:08:00] #         c:\Program Files (x86)\Microsoft Visual Studio 10.0\VC\INCLUDE\xtr1common(236) : see previous definition of '_Is_integral<char>'

[00:08:00] # c:\Program Files (x86)\Microsoft Visual Studio 10.0\VC\INCLUDE\ostream(545) : warning C4003: not enough actual parameters for macro 'write'
[00:08:00] # c:\Program Files (x86)\Microsoft Visual Studio 10.0\VC\INCLUDE\ostream(544) : error C2059: syntax error : ')'
[00:08:00] #         c:\Program Files (x86)\Microsoft Visual Studio 10.0\VC\INCLUDE\ostream(604) : see reference to class template instantiation 'std::basic_ostream<_Elem,_Traits>' being compiled
[00:08:00] # c:\Program Files (x86)\Microsoft Visual Studio 10.0\VC\INCLUDE\ostream(546) : error C2143: syntax error : missing ')' before '{'

@bulk88 can you check these errors? My win32 vm's just got bonkers with a new debian update and apparmor nonsense.

**Comment by rurban** [comment]
Perl_do_open is not propely exported as mathom

**Comment by rurban** [comment]
Closed with 940bb03228d8945b163d64ed7956dedb11cdee69

**Comment by rurban** [comment]
fix do_close clash for freebsd, clang++
added to v5.25.3c
